x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java

```diff
@@ -727,9 +727,9 @@ private void asyncNodeOperation(NodeTermsEnumRequest request, Task task, ActionL
                         false
                     )
                 );
-            }
-            if (canMatchShard(shardId, request) == false) {
-                // Permission denied or can't match, remove shardID from request
+                request.remove(shardId);
```

Comment:
I don't follow why we would do anything, especially modify the request, after returning a response to the caller. Intuitively, if request needs changing we would do that before returning a response, and possibly there would be a `return;` after the `listener.onResponse` call?